### ISSUE TITLE
[Nightly] History line not drawn properly (or not at all)

### ISSUE THREAD

**Issue by MagpieFourtyTwo** [cgeo/cgeo#8663]
**Describe the bug:**
History line is not drawn properly if map view is closed and re-opened again.

**To Reproduce:**
1. open map view (no matter if live or cache context, tested with GM and OSM v5)
2. move - around corners (to produce edges) ... and watch history line growing ...
3. leave map view (not matter if via switching display off or changing to cache detail page or to a different app)
4. move a little
5. re-open map view ...
6. History line will not be (re-)drawn like it was before, i. e. show the curves you've taken (and which have definitely been recorded in step 2.), but will be a straight bee's line back to a way older point (I would say, it's the one where you have been when you switched to map view the first time). BTW: Previously recorded parts of the history will neverhtheless be shown correctly.
7. move a little ... and watch history line growing again: newly recorded parts will be drawn properly again (while the line from step 6 will remain straight)
8. leave map like in step 3.
9. move a little 
10. re-open map view ...
11. newly recorded parts (step 7) will again not be shown, and there will again just be a straight line from your current position to the position mentioned in step 6.

Instead of steps 8. to 11. it sometimes is even sufficient to switch map products (e. g. from OSM to GM or vice versa).

**Expected behavior/state after performing these steps:**
History line should always display *all* recorded points.

**Version of c:geo used:**
2020.07.20-NB-33a57ef

**Is the problem reproducible:**
Yes

**Additional context:**
Although these newly recorded parts of the history line seem to have vanished, they will suddenly appear again on a fresh start of c:geo (you may even have to remove c:geo from "recent apps").

Moreover: Sometimes, when opening map view after some usage of c:geo, the entire history line will not be displayed at all - until you do a fresh restart, too (and sometimes just switching map context [e. g. from live to map context] may be sufficient) - but most regretfully I was not able to find a deterministic way to reproduce this ... perhaps someone else observed it, too ...?

**Comment by MagpieFourtyTwo** [comment]
@moving-bits wrote in #8664 :
> I adapted the algorithm for Mapsforge NewMap to that from Google Maps v2 - let's see if that's better (see #8665).

> Meanwhile I've debugged much deeper, and I think I have an idea what's happening. Trail gets truncated on startup only, but reloaded frequently on map restart/resume. So let's imagine you have accumulated A-B-C-D-E, but max length set to three, then on resume it will load A-B-C only (after #8649). So I think I will need to load in reverse order and max. length (giving E-D-C) and reversing that (to receive C-D-E). Will try that.

Wouldn't it be easier to limit in-memory recording to max length, too ...?  As meanwhile this limit can easily be set to a higher value anyway, there should be no reason anymore for a (seemingly) extended recording ... even more as it is temporary anyway.

Nevertheless, although I don't expect a change with regard to this topic (as GM did it wrong, too), I will re-check #8663 tomorrow with the adapted history line implementation from #8665

**Comment by moving-bits** [comment]
> Wouldn't it be easier to limit in-memory recording to max length, too ...? As meanwhile this limit can easily be set to a higher value anyway, there should be no reason anymore for a (seemingly) extended recording ... even more as it is temporary anyway.

This is already done, AFAIK, although not on every GPS signal.

> Nevertheless, although I don't expect a change with regard to this topic (as GM did it wrong, too), I will re-check #8663 tomorrow with the adapted history line implementation from #8665

I don't expect a change here regarding #8663, the mitigation with #8665 is for #8664 only.
Couldn't finish debugging #8663 yesterday, will continue today or tomorrow, as time permits.

**Comment by moving-bits** [comment]
@MagpieFourtyTwo 
Did you happen to have a chance to test if the fix works for you?

**Comment by MagpieFourtyTwo** [comment]
YESSIR!  ;)

Works surprisingly well!  ;)

No, seriously: Did not notice any peculiarities anymore.

Esp. the "leave map, re-enter map" stuff, where the most recent parts of the history line were displayed as a straight line instead of following the trail after re-entering, seems to work flawlessly now, moreover, the connection between history line and current position does not get lost anymore ... looks really good so far.

But, to be honest, had just four to five somewhat short test rides and walks so far, about half an hour each, and I still have this non-reproducable and non-deterministic "history line not shown at all" thingy in mind. So I will see on weekend how history behaves under more load ... but in fact I'm really optimistic - I really like how good c:geo currently works. ;)

**Comment by moving-bits** [comment]
> No, seriously: Did not notice any peculiarities anymore.

That's good news! Thanks for testing.
We can closing in on a beta then, IMHO.

> But, to be honest, had just four to five somewhat short test rides and walks so far, about half an hour each, and I still have this non-reproducable and non-deterministic "history line not shown at all" thingy in mind. So I will see on weekend how history behaves under more load ... but in fact I'm really optimistic - I really like how good c:geo currently works. ;)

Haven't seen that myself so far, but if you find anything reproducible for that, I would be happy to try a fix for it.

**Comment by MagpieFourtyTwo** [comment]
> We can closing in on a beta then, IMHO.

Agreed.
